Error correlator and troubleshoot webview

traeok opened the pull request:
## Proposed changes

- Implements an error correlation facility, allowing us and extenders to provide user-friendly summaries of encountered errors
  - We (as well as extenders) can also provide error-specific tips and external resources alongside the correlations to help guide users 😋 
  - Summaries are displayed in the initial notification if one was found; otherwise, the full error message is rendered in the dialog with some options
- Implements a troubleshoot webview, giving users additional information on the error
  - Error code, description, full error details
  - "Copy details" button to copy full error details & stack trace to clipboard
  - Tips (if available for the error correlation)
  - Resources (if available)

**If a correlation is found:**
- The error message summary is shown in the dialog. A "More info" option is presented.
- If "More info" is clicked, a secondary dialog is shown with the full error details and the "Show log" & "Troubleshoot" options mentioned below.

![image](https://github.com/user-attachments/assets/65c3cadc-6886-47d6-b4ff-4f9eaa01f9ff)

After clicking more info:
![image](https://github.com/user-attachments/assets/3ec18269-95d6-48da-ab69-69f8131db4c3)

**If a correlation is not found:**
- The full error details are shown in the dialog. It also offers a "Show log" and "Troubleshoot" option for diagnosing the error.

![image](https://github.com/user-attachments/assets/f33568a0-4de6-4add-acb6-ce495292e5bf)

**If "Troubleshoot" is clicked in the dialogs:**
- A troubleshoot webview is opened, displaying the error code, description and full error details - as well as tips and resources (if available)
- A "Copy details" button is available to copy the contents of the error message (details and stack trace)

<details>
<summary>Click to reveal troubleshoot screenshot</summary>
<img src="https://github.com/user-attachments/assets/1fab54b3-d641-465a-a5de-ac0b15563739" />
</details>

## Release Notes

<!-- Include the Milestone Number and a small description of your change that will be added to the changelog. -->
<!-- If there is a linked issue, it should have the same milestone as this PR. -->

Milestone: 3.1.0

Changelog:

**ZE:**

- Implemented more user-friendly error messages for API or network errors within Zowe Explorer. [#3243](https://github.com/zowe/zowe-explorer-vscode/pull/3243)
- Use the "Troubleshoot" option for certain errors to obtain additional context, tips, and resources for how to resolve the errors. [#3243](https://github.com/zowe/zowe-explorer-vscode/pull/3243)

**ZE API:**

- Leverage the new error correlation facility to provide user-friendly summaries of API and network errors. Extenders can also contribute to the correlator to provide human-readable translations of error messages, as well as tips and additional resources for how to resolve the error. [#3243](https://github.com/zowe/zowe-explorer-vscode/pull/3243)

## Types of changes

<!-- What types of changes does your code introduce to Zowe Explorer? Put an `x` in all boxes that apply. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

<!-- Put an `x` in all boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer. -->

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] All PR dependencies have been merged and published (if applicable)
- [x] A GIF or screenshot is included in the PR for visual changes
- [x] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):
](https://github.com/user-attachments/assets/1fab54b3-d641-465a-a5de-ac0b15563739)

traeok commented on packages/zowe-explorer/src/webviews/src/troubleshoot-error/App.tsx:
```diff
@@ -0,0 +1,46 @@
+  const [errorInfo, setErrorInfo] = useState<ErrorInfoProps>();
+
+  useEffect(() => {
+    window.addEventListener("message", (event) => {
```
There is no problem here - this check is already in place and the AI is merely suggesting to add a comment that would bloat the output in Developer Tools.

traeok commented:
Sorry for the early request to review - need to account for a couple more things before its ready:

- Allow extenders to contribute to the correlator
- Allow extenders to add custom resources to the correlator to display during Troubleshooting (stretch)

anaxceron commented on packages/zowe-explorer-api/CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ All notable changes to the "zowe-explorer-api" extension will be documented in t
 
 - Zowe Explorer now includes support for the [VS Code display languages](https://code.visualstudio.com/docs/getstarted/locales) French, German, Japanese, Portuguese, and Spanish. [#3239](https://github.com/zowe/zowe-explorer-vscode/pull/3239)
 - Localization of strings within the webviews. [#2983](https://github.com/zowe/zowe-explorer-vscode/issues/2983)
+- Implemented an error correlation facility for providing user-friendly summaries of API and network errors. Extenders can contribute to the correlator to provide human-readable translations of error messages, as well as tips and additional resources for how to resolve the error. [#3243](https://github.com/zowe/zowe-explorer-vscode/pull/3243)
```
Can we make it more action based?: "Apply (?) the new correlation facility to provide user-friendly summaries of API and network errors. Extenders can ... ."

anaxceron commented on packages/zowe-explorer/CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 
 - Update Zowe SDKs to `8.2.0` to get the latest enhancements from Imperative.
 - Added expired JSON web token detection for profiles in each tree view (Data Sets, USS, Jobs). When a user performs a search on a profile, they are prompted to log in if their token expired. [#3175](https://github.com/zowe/zowe-explorer-vscode/issues/3175)
+- Implemented more user-friendly error messages when encountering API or network errors within Zowe Explorer. [#3243](https://github.com/zowe/zowe-explorer-vscode/pull/3243)
```
A little clarity here, maybe?: "Implemented more user-friendly messages **for** API or network errors in Zowe Explorer."

anaxceron commented on packages/zowe-explorer/CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 - Update Zowe SDKs to `8.2.0` to get the latest enhancements from Imperative.
 - Added expired JSON web token detection for profiles in each tree view (Data Sets, USS, Jobs). When a user performs a search on a profile, they are prompted to log in if their token expired. [#3175](https://github.com/zowe/zowe-explorer-vscode/issues/3175)
+- Implemented more user-friendly error messages when encountering API or network errors within Zowe Explorer. [#3243](https://github.com/zowe/zowe-explorer-vscode/pull/3243)
+- Added a "Troubleshoot" option for certain errors to provide additional context, tips, and resources for how to resolve them. [#3243](https://github.com/zowe/zowe-explorer-vscode/pull/3243)
```
I'd try a more action-drive approach: "Use the 'Troubleshoot' option for certain errors to obtain additional context, tips, and resources for how to resolve the errors."

anaxceron reviewed (commented):
Made some suggestions for the changelogs.

anaxceron reviewed (approved):
Nice changelogs, thanks @traeok!

JillieBeanSim reviewed (commented):
OMG @traeok ! This is so cool!
I do see with the troubleshooting window the error details wasn't filled in, the rest looks great.  
Thanks for working on this
![Screenshot 2024-10-24 at 12 48 19 PM](https://github.com/user-attachments/assets/5ea0467e-d080-4525-9d07-5003a59da0f0)

traeok commented:
Thanks for the feedback Billie - I'm seeing a couple errors pop up when using virtual workspaces. Going to block this for now while I resolve this

zFernand0 commented on packages/zowe-explorer-api/src/utils/ErrorCorrelator.ts:
```diff
@@ -0,0 +1,360 @@
+            ErrorCorrelator.instance = new ErrorCorrelator();
+        }
+
+        return ErrorCorrelator.instance;
```
I know we've used `get` in front of our properties (`getApi...`, `getInstance`, ... ).
I'm curious if it makes sense to take advantage of getters (even though they can be a bit of a pain to mock during tests).

If we do decide to use getter, it may make the use a bit better for  extenders, i.e. they will be able to just do `ErrorCorrelator.instance` (similar to what Imperative does with many of its classes, like ImperativeConfig.instance and others)

traeok replied:
Not to play devil's advocate here because I agree that we should have getters with just the name `instance`. However for consistency's sake, I think it makes sense to keep it as `getInstance`, as we have other singletons in ZE and ZE API (table facilities) with the same pattern. We could address this across all our singletons in v4 😋

zFernand0 replied:
🙏 
yeah, I `getIt()` 😋 
Make sense to keep things consistent 🙏

zFernand0 replied:
Created:
- https://github.com/zowe/zowe-explorer-vscode/issues/3284

zFernand0 commented on packages/zowe-explorer/__tests__/__unit__/trees/dataset/ZoweDatasetNode.unit.test.ts:
```diff
@@ -520,7 +520,7 @@ describe("ZoweDatasetNode Unit Tests - Function node.openDs()", () => {
         }
 
-        expect(mocked(Gui.errorMessage)).toHaveBeenCalledWith("Error: testError");
+        expect(mocked(Gui.errorMessage)).toHaveBeenCalledWith("testError", { items: ["Show log", "Troubleshoot"] });
```
This a bit nuanced, but I'm guessing the removal of the word `Error: ` across the board came about when removing the `toString()` from errors?

Not saying I disagree with the change, but I'm curious to know what inspired the change 😋

traeok replied:
The "error" prefix seemed redundant since VS Code already renders it as an error message (red X beside notification), and we typically mention that there was a failure or error within the error text:

![image](https://github.com/user-attachments/assets/90b362cc-56f0-4681-bd5e-7f1dd4543f2e)

That said, I don't mind adding it back if the team prefers the prefix. This was just removed as a matter of preference 😁

zFernand0 replied:
I think it looks cleaner.

but in terms of the way we implement error handling... Having the `Error: ` prefix is kind of the lazy way to show errors (since I believe you can do an `error.toString()` inside a catch without an `instanceof Error` check) 😋

zFernand0 commented on packages/zowe-explorer/src/trees/shared/SharedInit.ts:
```diff
@@ -393,7 +406,13 @@ export class SharedInit {
         );
         for (const folder of newWorkspaces) {
-            await (folder.uri.scheme === ZoweScheme.DS ? DatasetFSProvider.instance : UssFSProvider.instance).remoteLookupForResource(folder.uri);
+            try {
+                await (folder.uri.scheme === ZoweScheme.DS ? DatasetFSProvider.instance : UssFSProvider.instance).remoteLookupForResource(folder.uri);
+            } catch (err) {
+                if (err instanceof Error) {
+                    ZoweLogger.error(err.message);
```
Thanks for catching this 🙏

zFernand0 reviewed (commented):
implementation LGTM so far 😋 
Have a bunch of dumb questions, but hopefully nothing tat requires much change (if any) 😅

zFernand0 commented on packages/zowe-explorer/src/trees/uss/UssFSProvider.ts:
```diff
@@ -120,23 +121,39 @@ export class UssFSProvider extends BaseProvider implements vscode.FileSystemProv
+            if (response.success && !keepRelative && response.apiResponse.items?.[0]?.mode?.startsWith("d") && !this.exists(uri)) {
+                await vscode.workspace.fs.createDirectory(uri);
+            }
+        } catch (err) {
+            if (err instanceof Error) {
```
still not a big fan of this, but I guess that's how TS want's people to handle it 😢 
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-4.html#defaulting-to-the-unknown-type-in-catch-variables---useunknownincatchvariables

Why not allow `catch(e: Error) { ... }` 💔

zFernand0 replied:
big sad 😅

zFernand0 commented on packages/zowe-explorer/src/utils/AuthUtils.ts:
```diff
@@ -11,84 +11,99 @@
 import { ZoweLogger } from "../tools/ZoweLogger";
 import { SharedTreeProviders } from "../trees/shared/SharedTreeProviders";
 
+export interface ErrorContext {
```
not sure we need to export it since we don't seems to use it anywhere else in the ZE package 😋

traeok replied:
ah, good point, thanks - will address in commit with other feedback addressed 🙂

zFernand0 reviewed (dismissed):
LGTM! 😋 

---

A couple of questions (which I'm not really asking for those to be implemented)

1. Is there a way to keep the notification around (maybe with some configuration setting) even after the user selects either "show log" or "Troubleshoot"?
2. Is it possible to scroll up on the log to where the error occurred when clicking `show log`?
3. Would it be possible to re-use (also could be configurable) the troubleshoot webview when viewing another error?

traeok commented on packages/zowe-explorer-api/src/fs/BaseProvider.ts:
```diff
@@ -417,6 +429,24 @@ export class BaseProvider {
+                    await opts.retry.fn(...(opts?.retry.args ?? []));
+                }
+            })
+            .catch(() => {
+                throw err;
```
Good call to add a log message to this catch block - we'll still need to throw the original error so that VS Code's file system handling is aware of the error that occurred in the filesystem.

traeok replied:
I now recall why I did not add an error message here - we do not expose the `ZoweLogger` to Zowe Explorer API. I still think we should add a log to this block, but not sure the best way to add it without adding a good chunk of code to expose it to the API (scope creep & more for reviewers to look over). 

Would you be opposed to adding a `console.log` call? I don't expect errors to be thrown from the correlator, but I added this catch block so that ESLint was satisfied with this use of a promise.

zFernand0 replied:
console.error might owkr for now 😋 
but I think it's be good to figure out how to do proper logging from the ZE APIs 😅 

Thanks for considering this 🙏

zFernand0 replied:
Created: 
- https://github.com/zowe/zowe-explorer-vscode/issues/3285

traeok commented on packages/zowe-explorer-api/src/fs/BaseProvider.ts:
```diff
@@ -15,6 +15,18 @@ import * as path from "path";
 import { ZosEncoding } from "../tree";
+import { ErrorCorrelator, ZoweExplorerApiType } from "../utils/ErrorCorrelator";
+
+export interface HandleErrorOpts {
```
The `fs/types` file is mostly for the file system types (data set / USS / jobs / generic entries) rather than interfaces, but I can see value in it being moved there for organization. 

I figured that this interface is only useful in the context of `BaseProvider`, so that you can grab everything you need for the base provider in one file 😋 if it doesn't cause any circular imports, I'll move it 👍

traeok commented on packages/zowe-explorer-api/src/tree/ZoweTreeNode.ts:
```diff
@@ -100,22 +100,9 @@ export class ZoweTreeNode extends vscode.TreeItem {
-        for (const child of this.children) {
-            (child as unknown as ZoweTreeNode).setProfileToChoice(aProfile, fsProvider);
-        }
+    public setProfileToChoice(aProfile: imperative.IProfileLoaded): void {
```
Ah, sorry this was left behind from testing and not related to the changes for the error correlator. I've restored the changes to this file in an upcoming commit. Thanks

traeok commented on packages/zowe-explorer-api/src/extend/IApiExplorerExtender.ts:
```diff
@@ -45,4 +46,11 @@ export interface IApiExplorerExtender {
+     * summaries of API or network errors. Also gives extenders the opportunity to
+     * provide tips or additional resources for errors.
+     */
+    getErrorCorrelator(): ErrorCorrelator;
```
Yes! Thanks for catching this 🙏

traeok commented on packages/zowe-explorer-api/CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ All notable changes to the "zowe-explorer-api" extension will be documented in t
 
 - Zowe Explorer now includes support for the [VS Code display languages](https://code.visualstudio.com/docs/getstarted/locales) French, German, Japanese, Portuguese, and Spanish. [#3239](https://github.com/zowe/zowe-explorer-vscode/pull/3239)
 - Localization of strings within the webviews. [#2983](https://github.com/zowe/zowe-explorer-vscode/issues/2983)
+- Leverage the new error correlation facility to provide user-friendly summaries of API and network errors. Extenders can also contribute to the correlator to provide human-readable translations of error messages, as well as tips and additional resources for how to resolve the error. [#3243](https://github.com/zowe/zowe-explorer-vscode/pull/3243)
```
Technically anyone using the Zowe Explorer API can leverage the error correlator. This was per Ana's suggestion to make changelog entries more of an "advertisement/announcement" for what devs can do rather than technical details of the change.

traeok commented on packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:
```diff
@@ -184,7 +184,8 @@ export class ProfilesCache {
                     // Step 3: Update allProfiles list
-                    tmpAllProfiles.push(profileFix);
+                    const existingProfile = this.allProfiles.find((tmpProf) => tmpProf.name === prof.profName && tmpProf.type === prof.profType);
+                    tmpAllProfiles.push(existingProfile ? Object.assign(existingProfile, profileFix) : profileFix);
```
Spread also works, but I think this is more explicit (thanks Timothy 😋)

traeok commented on packages/zowe-explorer-api/src/utils/ErrorCorrelator.ts:
```diff
@@ -0,0 +1,360 @@
+                    {
+                        errorCode: "500",
+                        matches: ["Client is not authorized for file access.", /An I\/O abend was trapped\.(.+?)\n(.+?)__code=0x0913/],
+                        summary: "Insufficient write permissions for this data set. The data set may be read-only or locked.",
```
This is a pain point I ran into when writing these error correlations. Since we don't have l10n set up for ZE API, we can't localize these strings. It'd be nice to add localization to the API, but since this would be a large, additional effort, I'll track this as an enhancement 🙂

zFernand0 replied:
Thanks for considering this 🙏

zFernand0 replied:
Created:
-  https://github.com/zowe/zowe-explorer-vscode/pull/3243#:~:text=ZE%2DAPI%3A%20Allow%20strings%20in%20the%20API%20to%20be%20localized%20%233283

traeok commented:
> LGTM! 😋
> 
> A couple of questions (which I'm not really asking for those to be implemented)
> 
> 1. Is there a way to keep the notification around (maybe with some configuration setting) even after the user selects either "show log" or "Troubleshoot"?
> 2. Is it possible to scroll up on the log to where the error occurred when clicking `show log`?
> 3. Would it be possible to re-use (also could be configurable) the troubleshoot webview when viewing another error?

1. Not that I'm aware of - however they are still accessible through the notification center: 
![image](https://github.com/user-attachments/assets/5db1307a-2b83-4bed-8112-b257c7a68717)
2. Unfortunately the output channel does not offer that capability.
3. Yes, anyone within ZE can create an error correlation and instantiate the troubleshoot view by constructing an instance of `TroubleshootError` with the extension context and error data as parameters.

traeok commented:
> Changes LGTM! 😋 Thanks for addressing the feedback 🙏
> 
> just one small comment about the optional v. required interface property 😋

@zFernand0 sorry, forgot to push that up. Should be good to go now.

zFernand0 reviewed (approved):
LGTM! 😋

zFernand0 commented:
> > LGTM! 😋
> > A couple of questions (which I'm not really asking for those to be implemented)
> > 
> > 1. Is there a way to keep the notification around (maybe with some configuration setting) even after the user selects either "show log" or "Troubleshoot"?
> > 2. Is it possible to scroll up on the log to where the error occurred when clicking `show log`?
> > 3. Would it be possible to re-use (also could be configurable) the troubleshoot webview when viewing another error?
> 
> 1. Are you asking if they will still persist after some time has elapsed? They will disappear at some point, but they are still accessible through the notification center:
>    ![image](https://private-user-images.githubusercontent.com/86500639/381132957-5db1307a-2b83-4bed-8112-b257c7a68717.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAzMDE0ODQsIm5iZiI6MTczMDMwMTE4NCwicGF0aCI6Ii84NjUwMDYzOS8zODExMzI5NTctNWRiMTMwN2EtMmI4My00YmVkLTgxMTItYjI1N2M3YTY4NzE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMwVDE1MTMwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4NjBlOTQ3NWMyOGVmMTY4MzQxODFmYWFjZWM1OTE5YjQ3ZGY0MGNmYmI3ODgzYmY3NjE1Y2I0ZWFmN2UxNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zWb5sqi4kB_q7qjvrhxvVg-khD1AKIZKEUQnNtHHN2k)
> 2. Unfortunately the output channel does not offer that capability.
> 3. Yes, anyone within ZE can create an error correlation and instantiate the troubleshoot view by constructing an instance of `TroubleshootError` with the extension context and error data as parameters.

1. This was kind os a silly request for the notification not to go away after opening the webview, but I don't think we have control over the notifications once the user clicks a button on them 😋 
2. bid sad 😢 
3. I meant for us to not continue splitting the view to the right, since you can have 2 or 3 webviews splits in the editor tab 😋

traeok commented:
> 3. I meant for us to not continue splitting the view to the right, since you can have 2 or 3 webviews splits in the editor tab 😋

Ah, gotcha - we can probably re-use the existing troubleshoot view if one is open, just may take some refactoring. Wouldn't mind looking into it if you'd prefer that behavior.

zFernand0 commented:
> > 3. I meant for us to not continue splitting the view to the right, since you can have 2 or 3 webviews splits in the editor tab 😋
> 
> Ah, gotcha - we can probably re-use the existing troubleshoot view if one is open, just may take some refactoring. Wouldn't mind looking into it if you'd prefer that behavior.

no need (for this PR)
I don't believe It's very common that people will need to open multiple webviews from a single error scenario 😋

zFernand0 reviewed (dismissed):
LGTM! 😋 
thanks for running the prepublish scripts 🙏

JillieBeanSim reviewed (approved):
LGTM! Thanks for working on this @traeok and for addressing feedback ❤️

zFernand0 reviewed (approved):
Changes LGTM! 😋 
left a small question/suggestion, but feel free to ignore if it doesn't make much sense 😋

traeok commented on packages/zowe-explorer/src/utils/AuthUtils.ts:
```diff
@@ -83,6 +83,7 @@ export class AuthUtils {
         const correlation = ErrorCorrelator.getInstance().correlateError(moreInfo?.apiType ?? ZoweExplorerApiType.All, errorDetails, {
             profileType: profile?.type,
             ...Object.keys(moreInfo).reduce((all, k) => (typeof moreInfo[k] === "string" ? { ...all, [k]: moreInfo[k] } : all), {}),
+            templateArgs: { ...(moreInfo.templateArgs ?? {}), profileName: profile?.name ?? "" },
```
The template args are essentially key/value pairs of strings, and in that scenario you would want to pass the profileName as a template arg rather than the profile, as these template args are passed right into Handlebars. If you have access to the profile at the time of the errorHandling call, it should be passed into `moreInfo` as the `profile` property.

As for the `?? {}` check, it is needed because `moreInfo` may be undefined - that said I do see a missing `?` before `.templateArgs`, so I'll add that.

t1m0thyj commented on packages/zowe-explorer/src/trees/uss/UssFSProvider.ts:
Is it intended behavior that I see 3 error messages when saving a USS file with invalid credentials? I would expect to see only 2 that are formatted consistently with the errors shown when saving a data set:
![image](https://github.com/user-attachments/assets/e46c689a-54ad-44c8-8245-05ca3b757d97)

traeok replied:
I will remove the last error message that you're seeing for auto-detect encoding as its normally done as a part of a separate operation (such as listing items in the tree or saving a file). The top 2 errors match what is shown for data sets.

traeok replied:
Resolved in a2f0307

t1m0thyj reviewed (dismissed):
LGTM, thanks @traeok! Left a few questions

traeok commented on packages/zowe-explorer/src/trees/dataset/DatasetFSProvider.ts:
```diff
@@ -543,6 +599,20 @@ export class DatasetFSProvider extends BaseProvider implements vscode.FileSystem
+                this._handleError(err, {
+                    additionalContext: vscode.l10n.t({
+                        message: "Failed to save {0}",
+                        args: [entry.metadata.path],
```
Fixed in a2f0307

traeok commented:
Hi all, sorry for the confusion but unfortunately the error in #3289 will not be resolved with this work.
Currently there are bugs with `promptCredentials` that is causing the profile to not propagate to the file system.

Once this is resolved, the "Try again" button in the editor will work as it will use the updated creds. We should not prompt here as extenders will use `readFile` to access data set and USS resources, so this would be a breaking change for extenders if we halted a `readFile` operation to prompt the user for new creds.

t1m0thyj reviewed (approved):
LGTM, thanks @traeok!

zFernand0 reviewed (approved):
LGTM!

zFernand0 reviewed (approved):
still LGTM! 😋
